allennlp_models/common/model_card.py

```diff
@@ -0,0 +1,267 @@
+    groups not present in the evaluation data?
+    """
+
+    caveats_and_recommendations: Optional[str] = None
+
+
+@dataclass(frozen=True)
+class ModelCard(ModelCardInfo):
+    """
+    The model card stores the recommended attributes for model reporting
+    as described in the paper [Model Cards for Model Reporting (Mitchell et al, 2019)]
+    (https://arxiv.org/pdf/1810.03993.pdf).
+    """
+
+    id: str
+    name: Optional[str] = None
+    archive_file: Optional[str] = None
+    overrides: Optional[Mapping] = None
+    model_details: Optional[ModelDetails] = None
+    intended_use: Optional[IntendedUse] = None
+    factors: Optional[Factors] = None
+    metrics: Optional[Metrics] = None
+    evaluation_data: Optional[EvaluationData] = None
+    training_data: Optional[TrainingData] = None
+    quantitative_analyses: Optional[QuantitativeAnalyses] = None
+    ethical_considerations: Optional[EthicalConsiderations] = None
+    caveats_and_recommendations: Optional[CaveatsAndRecommendations] = None
+
+    def to_dict(self):
+        info = {}
+        for key, val in self.__dict__.items():
+            if key != "id":
+                if isinstance(val, ModelCardInfo):
+                    info.update(val.to_dict())
+                else:
+                    info[key] = val
+        return info
+
+
+def add_pretrained_model(
+    id: str,
+    model_class: Optional[type] = None,
+    name: Optional[str] = None,
+    archive_file: Optional[str] = None,
+    overrides: Optional[Mapping] = None,
+    model_details: Optional[Union[str, ModelDetails]] = None,
+    intended_use: Optional[Union[str, IntendedUse]] = None,
+    factors: Optional[Union[str, Factors]] = None,
+    metrics: Optional[Union[str, Metrics]] = None,
+    evaluation_data: Optional[Union[str, EvaluationData]] = None,
+    training_data: Optional[Union[str, TrainingData]] = None,
+    quantitative_analyses: Optional[Union[str, QuantitativeAnalyses]] = None,
+    ethical_considerations: Optional[Union[str, EthicalConsiderations]] = None,
+    caveats_and_recommendations: Optional[Union[str, CaveatsAndRecommendations]] = None,
+) -> ModelCard:
+    """
+    Creates a `ModelCard` object and registers it
+    in the global dict of available pretrained models.
+    """
+    assert id
+    config = {}
+    config["id"] = id
+    if not model_class:
+        try:
+            model_class = Model.by_name(id)
+        except ConfigurationError:
+            logger.warning("{} is not a registered model.".format(id))
+
+    if model_class:
+        name = name or model_class.__name__
+        model_details = model_details or get_description(model_class)
+
+    if archive_file and not archive_file.startswith("https:"):
+        archive_file = os.path.join(STORAGE_LOCATION, archive_file)
+
+    config["name"] = name
+    config["archive_file"] = archive_file
+
+    config["model_details"] = ModelDetails.from_object(model_details)
```

Comment:
I haven't looked at this too closely, so feel free to ignore this if it doesn't make sense, but what's going on here reminds me a whole lot of what already happens in the `FromParams` logic. Is there an opportunity to reuse that here instead of writing this kind of code again?

Reply:
I believe the `FromParams` logic is more for constructing the model object given the archive file (correct me if I'm wrong). The `ModelCard` is supposed to be more for information-purpose, where `model_details` are to store meta-information about the model (when it was trained, who developed it, etc). I have included the `archive_file` because we want to pass all this information to the demo.

Reply:
`FromParams` is a general framework for constructing objects from JSON dictionaries.  The reason I thought this was similar was because it automatically handles the case where your type is `Union[str, SomeObject]`.  What you are doing here is constructing a `ModelCard` object while handling union types, and then adding it to the list of pretrained models.  You could imagine instead having the constructor be separate, and having `FromParams` worry about handling these union types, instead of duplicating that logic in all of these classes.

This would let you, e.g., add a pretrained model to our list by just writing a JSON file with model card info, including a path to the archive file, and then we load whatever is in some directory using `FromParams` and can display them wherever we want.  This also has the benefit of being able to handle uploading these JSON files to a server somewhere, without needing any code modifications.  We could someday support a fully-automated third-party model upload site.

Reply:
I think this case is different. `FromParams` treats a single string `s` the same as `{"type": s}`, but this takes the single string and treats it as the first parameter to the constructor.

Reply:
Regarding the decision to include the config in python files instead of JSON files, @dirkgr and I spoke about this earlier and decided that anyone updating the json file can also update the python file. The current advantage to keeping it in code is that we pull some of the missing information from the `Model` object without having to manually fill it in. I am open to making changes if we decide otherwise.

Reply:
@matt-gardner I went through the `FromParams` logic, and I agree with @dirkgr . Since we want to provide an easy way to include some, if not all of the model card information, the goal of the base class `ModelCardInfo` and its method `from_object` to map the provided str argument to the first argument of the constructor. If I want to use FromParams, I would still need to reimplement the `from_params` method to do this, and I do not think that class is providing any additional functionality for this use case. Note: I am looking at the way `FromParams` is used in `Model`, but perhaps there is a more generalized way to use it that I am not grasping.

Reply:
Just to be clear on what I'm suggesting:

```python
class ModelCard(FromParams):
    def __init__(
        self,
        model_class: Optional[type] = None,
        ...
        model_details: Optional[Union[str, ModelDetails]] = None,
        ...
    ):
        ...
        if isinstance(model_details, str):
            # Transparent logic for what happens when you're given a string, that can be documented
            # in this class's docstring.
            model_details = ModelDetails(description=str)
        self.model_details = model_details
        ...

def add_pretrained_model(model_card: ModelCard):
    # This method now looks kind of unnecessary
    PRETRAINED_MODELS[model_card.name] = model_card
    return model_card
```

And then instead of having a bunch of calls to `add_pretrained_model` scattered around the repo, we have a single directory that has json files describing model cards for each of our pretrained models, with a single method that finds all of those json files, and calls, essentially `add_pretrained_model(ModelCard.from_params(json_file))`.  Nothing else in your code needs to change, except that `ModelCardInfo` can go away, because it's no longer necessary.

The benefit over what you have here is that the constructor logic lives in one place, instead of two places.  If you want to add a parameter to `ModelCard`, with the code as it is now you have to modify both the `ModelCard` class and the `add_pretrained_model` method.  This is basically _exactly_ the situation we were in with our initial version of `from_params`, before we read type annotations to automatically discover and construct parameters.  It also seems conceptually simpler; it's not clear why the constructor for `ModelCard` should actually live inside of `add_pretrained_model`.  

One minor hiccup is that the data classes will also need to inherit from `FromParams` currently, though it seems easy enough to just make `FromParams` work automatically on data classes, as they will always have type annotations.